File: MonglePet/Overlay/PetMovementGeometry.swift
import Foundation

nonisolated struct PetMovementPoint: Equatable, Sendable {
    let x: Double
    let y: Double

    var isFinite: Bool {
        x.isFinite && y.isFinite
    }
}

nonisolated struct PetMovementSize: Equatable, Sendable {
    let width: Double
    let height: Double

    var isValid: Bool {
        width.isFinite && width > 0
            && height.isFinite && height > 0
    }
}

nonisolated struct PetMovementRect: Equatable, Sendable {
    let origin: PetMovementPoint
    let size: PetMovementSize

    init(x: Double, y: Double, width: Double, height: Double) {
        origin = PetMovementPoint(x: x, y: y)
        size = PetMovementSize(width: width, height: height)
    }

    var minX: Double { origin.x }
    var minY: Double { origin.y }
    var maxX: Double { origin.x + size.width }
    var maxY: Double { origin.y + size.height }
    var midX: Double { origin.x + (size.width / 2) }
    var midY: Double { origin.y + (size.height / 2) }
    var center: PetMovementPoint { PetMovementPoint(x: midX, y: midY) }

    var isValid: Bool {
        origin.isFinite && size.isValid && maxX.isFinite && maxY.isFinite
    }

    func contains(_ point: PetMovementPoint) -> Bool {
        isValid && point.isFinite
            && point.x >= minX && point.x <= maxX
            && point.y >= minY && point.y <= maxY
    }
}

nonisolated struct PetMovementScreen: Equatable, Sendable {
    let id: String
    let visibleFrame: PetMovementRect
}

nonisolated struct PetMovementWindow: Equatable, Sendable {
    let frame: PetMovementRect
}

nonisolated struct PetMovementRandomSample: Equatable, Sendable {
    let screen: Double
    let horizontal: Double
    let vertical: Double
}

nonisolated struct PetMovementOriginBounds: Equatable, Sendable {
    let minX: Double
    let maxX: Double
    let minY: Double
    let maxY: Double

    func contains(_ point: PetMovementPoint) -> Bool {
        point.isFinite
            && point.x >= minX && point.x <= maxX
            && point.y >= minY && point.y <= maxY
    }

    func clamped(_ point: PetMovementPoint) -> PetMovementPoint {
        PetMovementPoint(
            x: min(max(point.x, minX), maxX),
            y: min(max(point.y, minY), maxY)
        )
    }

    func point(horizontal: Double, vertical: Double) -> PetMovementPoint {
        let normalizedHorizontal = Self.normalizedUnitValue(horizontal)
        let normalizedVertical = Self.normalizedUnitValue(vertical)
        return PetMovementPoint(
            x: minX + ((maxX - minX) * normalizedHorizontal),
            y: minY + ((maxY - minY) * normalizedVertical)
        )
    }

    func intersection(
        with other: PetMovementOriginBounds
    ) -> PetMovementOriginBounds? {
        let intersectionMinX = max(minX, other.minX)
        let intersectionMaxX = min(maxX, other.maxX)
        let intersectionMinY = max(minY, other.minY)
        let intersectionMaxY = min(maxY, other.maxY)
        guard intersectionMinX <= intersectionMaxX,
              intersectionMinY <= intersectionMaxY else {
            return nil
        }
        return PetMovementOriginBounds(
            minX: intersectionMinX,
            maxX: intersectionMaxX,
            minY: intersectionMinY,
            maxY: intersectionMaxY
        )
    }

    private static func normalizedUnitValue(_ value: Double) -> Double {
        guard value.isFinite else {
            return 0
        }
        return min(max(value, 0), 1)
    }
}

nonisolated struct PetMovementAdvance: Equatable, Sendable {
    let origin: PetMovementPoint
    let didMove: Bool
    let hasArrived: Bool
}

nonisolated struct PetMovementScreenTransition: Equatable, Sendable {
    let sourceScreenID: String
    let targetScreenID: String
    let exitOrigin: PetMovementPoint
    let entryOrigin: PetMovementPoint
}

nonisolated struct PetMovementCursorRoute: Equatable, Sendable {
    let targetOrigin: PetMovementPoint
    let transition: PetMovementScreenTransition?
}

nonisolated enum PetMovementGeometry {
    static func safeOriginBounds(
        in visibleFrame: PetMovementRect,
        petSize: PetMovementSize,
        inset: Double
    ) -> PetMovementOriginBounds? {
        guard visibleFrame.isValid, petSize.isValid, inset.isFinite else {
            return nil
        }

        let safeInset = max(0, inset)
        let minimumX = visibleFrame.minX + safeInset
        let maximumX = visibleFrame.maxX - safeInset - petSize.width
        let minimumY = visibleFrame.minY + safeInset
        let maximumY = visibleFrame.maxY - safeInset - petSize.height

        let centeredX = visibleFrame.midX - (petSize.width / 2)
        let centeredY = visibleFrame.midY - (petSize.height / 2)
        return PetMovementOriginBounds(
            minX: minimumX <= maximumX ? minimumX : centeredX,
            maxX: minimumX <= maximumX ? maximumX : centeredX,
            minY: minimumY <= maximumY ? minimumY : centeredY,
            maxY: minimumY <= maximumY ? maximumY : centeredY
        )
    }

    static func cursorFollowingTargetOrigin(
        pointer: PetMovementPoint,
        currentOrigin: PetMovementPoint,
        petSize: PetMovementSize,
        cursorDistance: Double,
        screenInset: Double,
        screens: [PetMovementScreen]
    ) -> PetMovementPoint? {
        cursorFollowingRoute(
            pointer: pointer,
            currentOrigin: currentOrigin,
            petSize: petSize,
            cursorDistance: cursorDistance,
            screenInset: screenInset,
            screens: screens
        )?.targetOrigin
    }

    static func cursorFollowingRoute(
        pointer: PetMovementPoint,
        currentOrigin: PetMovementPoint,
        petSize: PetMovementSize,
        cursorDistance: Double,
        screenInset: Double,
        screens: [PetMovementScreen]
    ) -> PetMovementCursorRoute? {
        guard pointer.isFinite, currentOrigin.isFinite, petSize.isValid,
              cursorDistance.isFinite, cursorDistance >= 0,
              let targetScreen = screen(
                  containingOrNearestTo: pointer,
                  in: screens
              ),
              let bounds = safeOriginBounds(
                  in: targetScreen.visibleFrame,
                  petSize: petSize,
                  inset: screenInset
              ) else {
            return nil
        }

        let currentCenter = PetMovementPoint(
            x: currentOrigin.x + (petSize.width / 2),
            y: currentOrigin.y + (petSize.height / 2)
        )
        let deltaX = pointer.x - currentCenter.x
        let deltaY = pointer.y - currentCenter.y
        let distance = hypot(deltaX, deltaY)

        let targetCenter: PetMovementPoint
        if distance > 0 {
            targetCenter = PetMovementPoint(
                x: pointer.x - ((deltaX / distance) * cursorDistance),
                y: pointer.y - ((deltaY / distance) * cursorDistance)
            )
        } else {
            targetCenter = pointer
        }

        let targetOrigin = bounds.clamped(
            PetMovementPoint(
                x: targetCenter.x - (petSize.width / 2),
                y: targetCenter.y - (petSize.height / 2)
            )
        )
        guard
            let sourceScreen = screen(
                containingOrNearestTo: currentCenter,
                in: screens
            ),
            sourceScreen.id != targetScreen.id,
            let sourceBounds = safeOriginBounds(
                in: sourceScreen.visibleFrame,
                petSize: petSize,
                inset: 0
            ),
            let targetBounds = safeOriginBounds(
                in: targetScreen.visibleFrame,
                petSize: petSize,
                inset: 0
            )
        else {
            return PetMovementCursorRoute(
                targetOrigin: targetOrigin,
                transition: nil
            )
        }

        let horizontalTransition = closestTransitionCoordinates(
            sourceMinimum: sourceBounds.minX,
            sourceMaximum: sourceBounds.maxX,
            targetMinimum: targetBounds.minX,
            targetMaximum: targetBounds.maxX,
            preferred: currentOrigin.x
        )
        let verticalTransition = closestTransitionCoordinates(
            sourceMinimum: sourceBounds.minY,
            sourceMaximum: sourceBounds.maxY,
            targetMinimum: targetBounds.minY,
            targetMaximum: targetBounds.maxY,
            preferred: currentOrigin.y
        )
        return PetMovementCursorRoute(
            targetOrigin: targetOrigin,
            transition: PetMovementScreenTransition(
                sourceScreenID: sourceScreen.id,
                targetScreenID: targetScreen.id,
                exitOrigin: PetMovementPoint(
                    x: horizontalTransition.source,
                    y: verticalTransition.source
                ),
                entryOrigin: PetMovementPoint(
                    x: horizontalTransition.target,
                    y: verticalTransition.target
                )
            )
        )
    }

    static func freeRoamingTargetOrigin(
        screens: [PetMovementScreen],
        petSize: PetMovementSize,
        screenInset: Double,
        preferredWindow: PetMovementWindow?,
        sample: PetMovementRandomSample
    ) -> PetMovementPoint? {
        let validScreens = screens.filter { $0.visibleFrame.isValid }
        guard !validScreens.isEmpty, petSize.isValid else {
            return nil
        }

        if let preferredWindow,
           let preferredBounds = preferredOriginBounds(
               for: preferredWindow,
               screens: validScreens,
               petSize: petSize,
               screenInset: screenInset
           ) {
            return preferredBounds.point(
                horizontal: sample.horizontal,
                vertical: sample.vertical
            )
        }

        let screenIndex = randomIndex(sample.screen, count: validScreens.count)
        guard let bounds = safeOriginBounds(
            in: validScreens[screenIndex].visibleFrame,
            petSize: petSize,
            inset: screenInset
        ) else {
            return nil
        }
        return bounds.point(
            horizontal: sample.horizontal,
            vertical: sample.vertical
        )
    }

    static func advance(
        from currentOrigin: PetMovementPoint,
        toward targetOrigin: PetMovementPoint,
        speed: Double,
        elapsedSeconds: Double,
        stopRadius: Double
    ) -> PetMovementAdvance {
        guard currentOrigin.isFinite, targetOrigin.isFinite,
              speed.isFinite, speed > 0,
              elapsedSeconds.isFinite, elapsedSeconds > 0,
              stopRadius.isFinite, stopRadius >= 0 else {
            return PetMovementAdvance(
                origin: currentOrigin,
                didMove: false,
                hasArrived: false
            )
        }

        let deltaX = targetOrigin.x - currentOrigin.x
        let deltaY = targetOrigin.y - currentOrigin.y
        let distance = hypot(deltaX, deltaY)
        guard distance > stopRadius else {
            return PetMovementAdvance(
                origin: currentOrigin,
                didMove: false,
                hasArrived: true
            )
        }

        let travelDistance = speed * elapsedSeconds
        guard travelDistance.isFinite, travelDistance > 0 else {
            return PetMovementAdvance(
                origin: currentOrigin,
                didMove: false,
                hasArrived: false
            )
        }

        if travelDistance >= distance {
            return PetMovementAdvance(
                origin: targetOrigin,
                didMove: true,
                hasArrived: true
            )
        }

        let progress = travelDistance / distance
        let nextOrigin = PetMovementPoint(
            x: currentOrigin.x + (deltaX * progress),
            y: currentOrigin.y + (deltaY * progress)
        )
        return PetMovementAdvance(
            origin: nextOrigin,
            didMove: true,
            hasArrived: (distance - travelDistance) <= stopRadius
        )
    }

    private static func screen(
        containingOrNearestTo point: PetMovementPoint,
        in screens: [PetMovementScreen]
    ) -> PetMovementScreen? {
        let validScreens = screens.filter { $0.visibleFrame.isValid }
        if let containingScreen = validScreens.first(where: {
            $0.visibleFrame.contains(point)
        }) {
            return containingScreen
        }
        return validScreens.min { lhs, rhs in
            squaredDistance(from: point, to: lhs.visibleFrame)
                < squaredDistance(from: point, to: rhs.visibleFrame)
        }
    }

    private static func preferredOriginBounds(
        for window: PetMovementWindow,
        screens: [PetMovementScreen],
        petSize: PetMovementSize,
        screenInset: Double
    ) -> PetMovementOriginBounds? {
        guard window.frame.isValid,
              window.frame.size.width >= petSize.width,
              window.frame.size.height >= petSize.height,
              let screen = screens.max(by: { lhs, rhs in
                  intersectionArea(lhs.visibleFrame, window.frame)
                      < intersectionArea(rhs.visibleFrame, window.frame)
              }),
              intersectionArea(screen.visibleFrame, window.frame) > 0,
              let screenBounds = safeOriginBounds(
                  in: screen.visibleFrame,
                  petSize: petSize,
                  inset: screenInset
              ),
              let windowBounds = safeOriginBounds(
                  in: window.frame,
                  petSize: petSize,
                  inset: 0
              ) else {
            return nil
        }
        return screenBounds.intersection(with: windowBounds)
    }

    private static func randomIndex(_ value: Double, count: Int) -> Int {
        guard value.isFinite else {
            return 0
        }
        let normalized = min(max(value, 0), 1)
        return min(Int(normalized * Double(count)), count - 1)
    }

    private static func closestTransitionCoordinates(
        sourceMinimum: Double,
        sourceMaximum: Double,
        targetMinimum: Double,
        targetMaximum: Double,
        preferred: Double
    ) -> (source: Double, target: Double) {
        if sourceMaximum < targetMinimum {
            return (sourceMaximum, targetMinimum)
        }
        if targetMaximum < sourceMinimum {
            return (sourceMinimum, targetMaximum)
        }

        let sharedMinimum = max(sourceMinimum, targetMinimum)
        let sharedMaximum = min(sourceMaximum, targetMaximum)
        let sharedCoordinate = min(
            max(preferred, sharedMinimum),
            sharedMaximum
        )
        return (sharedCoordinate, sharedCoordinate)
    }

    private static func squaredDistance(
        from point: PetMovementPoint,
        to rect: PetMovementRect
    ) -> Double {
        let nearestX = min(max(point.x, rect.minX), rect.maxX)
        let nearestY = min(max(point.y, rect.minY), rect.maxY)
        let deltaX = point.x - nearestX
        let deltaY = point.y - nearestY
        return (deltaX * deltaX) + (deltaY * deltaY)
    }

    private static func intersectionArea(
        _ lhs: PetMovementRect,
        _ rhs: PetMovementRect
    ) -> Double {
        let width = max(0, min(lhs.maxX, rhs.maxX) - max(lhs.minX, rhs.minX))
        let height = max(0, min(lhs.maxY, rhs.maxY) - max(lhs.minY, rhs.minY))
        return width * height
    }
}

File: apps/windows/src/MonglePet.Core/Movement/PetMovementGeometry.cs
namespace MonglePet.Core.Movement;

public readonly record struct MovementPoint(double X, double Y)
{
    public bool IsFinite => double.IsFinite(X) && double.IsFinite(Y);
}

public readonly record struct MovementSize(double Width, double Height)
{
    public bool IsValid =>
        double.IsFinite(Width) && Width > 0 &&
        double.IsFinite(Height) && Height > 0;
}

public readonly record struct MovementRect(double X, double Y, double Width, double Height)
{
    public double Left => X;
    public double Top => Y;
    public double Right => X + Width;
    public double Bottom => Y + Height;
    public double CenterX => X + (Width / 2);
    public double CenterY => Y + (Height / 2);
    public bool IsValid =>
        double.IsFinite(X) && double.IsFinite(Y) &&
        double.IsFinite(Width) && Width > 0 &&
        double.IsFinite(Height) && Height > 0 &&
        double.IsFinite(Right) && double.IsFinite(Bottom);

    public bool Contains(MovementPoint point) =>
        IsValid && point.IsFinite &&
        point.X >= Left && point.X <= Right &&
        point.Y >= Top && point.Y <= Bottom;
}

public sealed record MovementScreen(string Identifier, MovementRect WorkArea);

public readonly record struct MovementOriginBounds(
    double MinimumX,
    double MaximumX,
    double MinimumY,
    double MaximumY)
{
    public MovementPoint Clamp(MovementPoint point) => new(
        Math.Clamp(point.X, MinimumX, MaximumX),
        Math.Clamp(point.Y, MinimumY, MaximumY));

    public MovementPoint Point(double horizontal, double vertical) => new(
        MinimumX + ((MaximumX - MinimumX) * Unit(horizontal)),
        MinimumY + ((MaximumY - MinimumY) * Unit(vertical)));

    public MovementOriginBounds? Intersection(MovementOriginBounds other)
    {
        double minimumX = Math.Max(MinimumX, other.MinimumX);
        double maximumX = Math.Min(MaximumX, other.MaximumX);
        double minimumY = Math.Max(MinimumY, other.MinimumY);
        double maximumY = Math.Min(MaximumY, other.MaximumY);
        return minimumX <= maximumX && minimumY <= maximumY
            ? new MovementOriginBounds(minimumX, maximumX, minimumY, maximumY)
            : null;
    }

    private static double Unit(double value) =>
        double.IsFinite(value) ? Math.Clamp(value, 0, 1) : 0;
}

public readonly record struct MovementAdvance(
    MovementPoint Origin,
    bool DidMove,
    bool HasArrived);

public readonly record struct MovementRandomSample(
    double Screen,
    double Horizontal,
    double Vertical);

public enum MovementDirection
{
    Left,
    Right,
    Up,
    Down,
    UpLeft,
    UpRight,
    DownLeft,
    DownRight,
}

public static class PetMovementGeometry
{
    public const double DefaultScreenInset = 8;

    public static MovementOriginBounds? SafeOriginBounds(
        MovementRect workArea,
        MovementSize petSize,
        double inset = DefaultScreenInset)
    {
        if (!workArea.IsValid || !petSize.IsValid || !double.IsFinite(inset))
        {
            return null;
        }

        double safeInset = Math.Max(0, inset);
        double minimumX = workArea.Left + safeInset;
        double maximumX = workArea.Right - safeInset - petSize.Width;
        double minimumY = workArea.Top + safeInset;
        double maximumY = workArea.Bottom - safeInset - petSize.Height;
        double centeredX = workArea.CenterX - (petSize.Width / 2);
        double centeredY = workArea.CenterY - (petSize.Height / 2);
        return new MovementOriginBounds(
            minimumX <= maximumX ? minimumX : centeredX,
            minimumX <= maximumX ? maximumX : centeredX,
            minimumY <= maximumY ? minimumY : centeredY,
            minimumY <= maximumY ? maximumY : centeredY);
    }

    public static MovementPoint? ClampToNearestScreen(
        MovementPoint requestedOrigin,
        MovementSize petSize,
        IReadOnlyList<MovementScreen> screens,
        double inset = DefaultScreenInset)
    {
        MovementScreen? screen = ScreenContainingOrNearest(
            new MovementPoint(
                requestedOrigin.X + (petSize.Width / 2),
                requestedOrigin.Y + (petSize.Height / 2)),
            screens);
        MovementOriginBounds? bounds = screen is null
            ? null
            : SafeOriginBounds(screen.WorkArea, petSize, inset);
        return bounds?.Clamp(requestedOrigin);
    }

    public static MovementPoint? CursorFollowingTarget(
        MovementPoint pointer,
        MovementPoint currentOrigin,
        MovementSize petSize,
        double cursorDistance,
        IReadOnlyList<MovementScreen> screens,
        double inset = DefaultScreenInset)
    {
        if (!pointer.IsFinite || !currentOrigin.IsFinite || !petSize.IsValid ||
            !double.IsFinite(cursorDistance) || cursorDistance < 0)
        {
            return null;
        }

        MovementScreen? screen = ScreenContainingOrNearest(pointer, screens);
        MovementOriginBounds? bounds = screen is null
            ? null
            : SafeOriginBounds(screen.WorkArea, petSize, inset);
        if (bounds is null)
        {
            return null;
        }

        double centerX = currentOrigin.X + (petSize.Width / 2);
        double centerY = currentOrigin.Y + (petSize.Height / 2);
        double deltaX = pointer.X - centerX;
        double deltaY = pointer.Y - centerY;
        double distance = Math.Sqrt((deltaX * deltaX) + (deltaY * deltaY));
        double targetCenterX = pointer.X;
        double targetCenterY = pointer.Y;
        if (distance > 0.0001)
        {
            targetCenterX -= (deltaX / distance) * cursorDistance;
            targetCenterY -= (deltaY / distance) * cursorDistance;
        }

        return bounds.Value.Clamp(new MovementPoint(
            targetCenterX - (petSize.Width / 2),
            targetCenterY - (petSize.Height / 2)));
    }

    public static MovementPoint? FreeRoamingTarget(
        IReadOnlyList<MovementScreen> screens,
        MovementSize petSize,
        MovementRandomSample sample,
        double inset = DefaultScreenInset,
        MovementRect? preferredWindow = null)
    {
        MovementScreen[] valid = screens.Where(screen => screen.WorkArea.IsValid).ToArray();
        if (valid.Length == 0 || !petSize.IsValid)
        {
            return null;
        }

        if (preferredWindow is { } window &&
            PreferredOriginBounds(window, valid, petSize, inset) is { } preferredBounds)
        {
            return preferredBounds.Point(sample.Horizontal, sample.Vertical);
        }

        int index = Math.Min(
            valid.Length - 1,
            (int)Math.Floor(Unit(sample.Screen) * valid.Length));
        MovementOriginBounds? bounds = SafeOriginBounds(valid[index].WorkArea, petSize, inset);
        return bounds?.Point(sample.Horizontal, sample.Vertical);
    }

    public static double? DistanceFromPointerToPet(
        MovementPoint pointer,
        MovementPoint petOrigin,
        MovementSize petSize)
    {
        if (!pointer.IsFinite || !petOrigin.IsFinite || !petSize.IsValid)
        {
            return null;
        }

        double nearestX = Math.Clamp(pointer.X, petOrigin.X, petOrigin.X + petSize.Width);
        double nearestY = Math.Clamp(pointer.Y, petOrigin.Y, petOrigin.Y + petSize.Height);
        double deltaX = pointer.X - nearestX;
        double deltaY = pointer.Y - nearestY;
        return Math.Sqrt((deltaX * deltaX) + (deltaY * deltaY));
    }

    public static MovementPoint? CursorAvoidingTarget(
        MovementPoint pointer,
        MovementPoint currentOrigin,
        MovementSize petSize,
        double safeDistance,
        IReadOnlyList<MovementScreen> screens,
        double inset = DefaultScreenInset)
    {
        if (!pointer.IsFinite || !currentOrigin.IsFinite || !petSize.IsValid ||
            !double.IsFinite(safeDistance) || safeDistance < 0)
        {
            return null;
        }

        double centerX = currentOrigin.X + (petSize.Width / 2);
        double centerY = currentOrigin.Y + (petSize.Height / 2);
        double awayX = centerX - pointer.X;
        double awayY = centerY - pointer.Y;
        double magnitude = Math.Sqrt((awayX * awayX) + (awayY * awayY));
        if (magnitude <= 0.0001)
        {
            awayX = 1;
            awayY = 0;
            magnitude = 1;
        }

        double travel = safeDistance + Math.Max(petSize.Width, petSize.Height);
        var desired = new MovementPoint(
            currentOrigin.X + ((awayX / magnitude) * travel),
            currentOrigin.Y + ((awayY / magnitude) * travel));
        var candidates = new List<MovementPoint>();
        foreach (MovementScreen screen in screens.Where(value => value.WorkArea.IsValid))
        {
            if (SafeOriginBounds(screen.WorkArea, petSize, inset) is not { } bounds)
            {
                continue;
            }
            MovementPoint direct = bounds.Clamp(desired);
            candidates.Add(direct);
            candidates.Add(bounds.Point(0, 0));
            candidates.Add(bounds.Point(0, 1));
            candidates.Add(bounds.Point(1, 0));
            candidates.Add(bounds.Point(1, 1));
        }

        if (candidates.Count == 0)
        {
            return null;
        }

        double unitX = awayX / magnitude;
        double unitY = awayY / magnitude;
        MovementPoint[] awayCandidates = candidates
            .Where(candidate =>
                ((candidate.X - currentOrigin.X) * unitX) +
                ((candidate.Y - currentOrigin.Y) * unitY) >= -0.0001)
            .Distinct()
            .ToArray();
        if (awayCandidates.Length == 0)
        {
            awayCandidates = candidates.Distinct().ToArray();
        }

        MovementPoint[] safeCandidates = awayCandidates.Where(candidate =>
            DistanceFromPointerToPet(pointer, candidate, petSize) >= safeDistance).ToArray();
        if (safeCandidates.Length > 0)
        {
            return safeCandidates.MinBy(candidate => SquaredDistance(candidate, currentOrigin));
        }
        return awayCandidates.MaxBy(candidate =>
            DistanceFromPointerToPet(pointer, candidate, petSize) ?? 0);
    }

    public static bool ShouldRefreshCursorAvoidingTarget(
        MovementPoint pointer,
        MovementPoint? pointerAnchor,
        bool hasTarget,
        double minimumPointerTravel = 24)
    {
        if (!pointer.IsFinite ||
            !double.IsFinite(minimumPointerTravel) || minimumPointerTravel < 0)
        {
            return true;
        }
        if (!hasTarget || pointerAnchor is not { IsFinite: true } anchor)
        {
            return true;
        }

        return SquaredDistance(pointer, anchor) >=
            minimumPointerTravel * minimumPointerTravel;
    }

    public static MovementAdvance Advance(
        MovementPoint currentOrigin,
        MovementPoint targetOrigin,
        double speed,
        double elapsedSeconds,
        double stopRadius)
    {
        if (!currentOrigin.IsFinite || !targetOrigin.IsFinite ||
            !double.IsFinite(speed) || speed <= 0 ||
            !double.IsFinite(elapsedSeconds) || elapsedSeconds <= 0 ||
            !double.IsFinite(stopRadius) || stopRadius < 0)
        {
            return new MovementAdvance(currentOrigin, false, false);
        }

        double deltaX = targetOrigin.X - currentOrigin.X;
        double deltaY = targetOrigin.Y - currentOrigin.Y;
        double distance = Math.Sqrt((deltaX * deltaX) + (deltaY * deltaY));
        if (distance <= stopRadius)
        {
            return new MovementAdvance(currentOrigin, false, true);
        }

        double travel = speed * elapsedSeconds;
        if (!double.IsFinite(travel) || travel <= 0)
        {
            return new MovementAdvance(currentOrigin, false, false);
        }
        if (travel >= distance)
        {
            return new MovementAdvance(targetOrigin, true, true);
        }

        double progress = travel / distance;
        return new MovementAdvance(
            new MovementPoint(
                currentOrigin.X + (deltaX * progress),
                currentOrigin.Y + (deltaY * progress)),
            true,
            (distance - travel) <= stopRadius);
    }

    public static MovementDirection? Direction(
        double deltaX,
        double deltaY,
        bool usesDiagonals,
        MovementDirection? previousDirection = null,
        double hysteresisDegrees = 8)
    {
        if (!double.IsFinite(deltaX) || !double.IsFinite(deltaY) ||
            Math.Abs(deltaX) + Math.Abs(deltaY) <= 0.0001)
        {
            return null;
        }

        MovementDirection classified;
        if (!usesDiagonals)
        {
            classified = Math.Abs(deltaX) >= Math.Abs(deltaY)
                ? deltaX < 0 ? MovementDirection.Left : MovementDirection.Right
                : deltaY < 0 ? MovementDirection.Up : MovementDirection.Down;
        }
        else
        {
            double angle = Math.Atan2(deltaY, deltaX) * 180 / Math.PI;
            classified = angle switch
            {
                >= -22.5 and < 22.5 => MovementDirection.Right,
                >= 22.5 and < 67.5 => MovementDirection.DownRight,
                >= 67.5 and < 112.5 => MovementDirection.Down,
                >= 112.5 and < 157.5 => MovementDirection.DownLeft,
                >= 157.5 or < -157.5 => MovementDirection.Left,
                >= -157.5 and < -112.5 => MovementDirection.UpLeft,
                >= -112.5 and < -67.5 => MovementDirection.Up,
                _ => MovementDirection.UpRight,
            };
        }

        if (previousDirection is not { } previous ||
            (!usesDiagonals && IsDiagonal(previous)) ||
            !double.IsFinite(hysteresisDegrees) ||
            hysteresisDegrees <= 0)
        {
            return classified;
        }

        double currentAngle = Math.Atan2(deltaY, deltaX) * 180 / Math.PI;
        double halfSector = usesDiagonals ? 22.5 : 45;
        return AngularDistance(currentAngle, DirectionAngle(previous)) <=
            halfSector + hysteresisDegrees
                ? previous
                : classified;
    }

    public static IReadOnlyList<MovementDirection> CompatibleDirections(
        double deltaX,
        double deltaY,
        bool usesDiagonals,
        MovementDirection? previousDirection = null,
        double hysteresisDegrees = 8)
    {
        MovementDirection? exact = Direction(
            deltaX,
            deltaY,
            usesDiagonals,
            previousDirection,
            hysteresisDegrees);
        if (exact is null)
        {
            return Array.Empty<MovementDirection>();
        }

        double distance = Math.Sqrt((deltaX * deltaX) + (deltaY * deltaY));
        if (!double.IsFinite(distance) || distance <= 0.0001)
        {
            return Array.Empty<MovementDirection>();
        }

        MovementDirection[] stableOrder = usesDiagonals
            ? [
                MovementDirection.Left,
                MovementDirection.Right,
                MovementDirection.Up,
                MovementDirection.Down,
                MovementDirection.UpLeft,
                MovementDirection.UpRight,
                MovementDirection.DownLeft,
                MovementDirection.DownRight,
            ]
            : [
                MovementDirection.Left,
                MovementDirection.Right,
                MovementDirection.Up,
                MovementDirection.Down,
            ];
        return stableOrder
            .Select((direction, index) =>
            {
                (double X, double Y) vector = DirectionVector(direction);
                double similarity = ((deltaX / distance) * vector.X) +
                    ((deltaY / distance) * vector.Y);
                bool opposesX = Math.Abs(deltaX / distance) > 0.05 &&
                    vector.X != 0 && Math.Sign(vector.X) != Math.Sign(deltaX);
                bool opposesY = Math.Abs(deltaY / distance) > 0.05 &&
                    vector.Y != 0 && Math.Sign(vector.Y) != Math.Sign(deltaY);
                return (Direction: direction, Similarity: similarity, Index: index,
                    IsCompatible: similarity > 0.05 && !opposesX && !opposesY);
            })
            .Where(item => item.IsCompatible)
            .OrderByDescending(item => item.Direction == exact)
            .ThenByDescending(item => item.Similarity)
            .ThenBy(item => item.Index)
            .Select(item => item.Direction)
            .ToArray();
    }

    private static bool IsDiagonal(MovementDirection direction) => direction is
        MovementDirection.UpLeft or MovementDirection.UpRight or
        MovementDirection.DownLeft or MovementDirection.DownRight;

    private static double DirectionAngle(MovementDirection direction) => direction switch
    {
        MovementDirection.Right => 0,
        MovementDirection.DownRight => 45,
        MovementDirection.Down => 90,
        MovementDirection.DownLeft => 135,
        MovementDirection.Left => 180,
        MovementDirection.UpLeft => -135,
        MovementDirection.Up => -90,
        MovementDirection.UpRight => -45,
        _ => 0,
    };

    private static (double X, double Y) DirectionVector(MovementDirection direction)
    {
        const double diagonal = 0.7071067811865476;
        return direction switch
        {
            MovementDirection.Left => (-1, 0),
            MovementDirection.Right => (1, 0),
            MovementDirection.Up => (0, -1),
            MovementDirection.Down => (0, 1),
            MovementDirection.UpLeft => (-diagonal, -diagonal),
            MovementDirection.UpRight => (diagonal, -diagonal),
            MovementDirection.DownLeft => (-diagonal, diagonal),
            MovementDirection.DownRight => (diagonal, diagonal),
            _ => (0, 0),
        };
    }

    private static double AngularDistance(double left, double right)
    {
        double delta = Math.Abs(left - right) % 360;
        return delta > 180 ? 360 - delta : delta;
    }

    private static MovementScreen? ScreenContainingOrNearest(
        MovementPoint point,
        IReadOnlyList<MovementScreen> screens)
    {
        MovementScreen[] valid = screens.Where(screen => screen.WorkArea.IsValid).ToArray();
        return valid.FirstOrDefault(screen => screen.WorkArea.Contains(point))
            ?? valid.MinBy(screen => SquaredDistanceToRect(point, screen.WorkArea));
    }

    private static MovementOriginBounds? PreferredOriginBounds(
        MovementRect window,
        IReadOnlyList<MovementScreen> screens,
        MovementSize petSize,
        double inset)
    {
        if (!window.IsValid ||
            window.Width < petSize.Width ||
            window.Height < petSize.Height)
        {
            return null;
        }

        MovementScreen? screen = screens
            .Where(value => value.WorkArea.IsValid)
            .MaxBy(value => IntersectionArea(value.WorkArea, window));
        if (screen is null || IntersectionArea(screen.WorkArea, window) <= 0 ||
            SafeOriginBounds(screen.WorkArea, petSize, inset) is not { } screenBounds ||
            SafeOriginBounds(window, petSize, 0) is not { } windowBounds)
        {
            return null;
        }
        return screenBounds.Intersection(windowBounds);
    }

    private static double IntersectionArea(MovementRect left, MovementRect right)
    {
        double width = Math.Max(0, Math.Min(left.Right, right.Right) -
            Math.Max(left.Left, right.Left));
        double height = Math.Max(0, Math.Min(left.Bottom, right.Bottom) -
            Math.Max(left.Top, right.Top));
        return width * height;
    }

    private static double SquaredDistanceToRect(MovementPoint point, MovementRect rect)
    {
        double x = Math.Clamp(point.X, rect.Left, rect.Right);
        double y = Math.Clamp(point.Y, rect.Top, rect.Bottom);
        return SquaredDistance(point, new MovementPoint(x, y));
    }

    private static double SquaredDistance(MovementPoint left, MovementPoint right)
    {
        double x = left.X - right.X;
        double y = left.Y - right.Y;
        return (x * x) + (y * y);
    }

    private static double Unit(double value) =>
        double.IsFinite(value) ? Math.Clamp(value, 0, 0.999999999) : 0;
}
